### PR TITLE
Write proper error messages for Get-Command ' '

### LINKED ISSUE #13342: Get-Command fails with unexpected error when given a blank command-name argument
Note: This is definitely a minor bug that won't arise too often in the real world.

Passing an all-whitespace `-Name` argument to `Get-Command` results in unexpected error message `Get-Command: Index was outside the bounds of the array.`

## Steps to reproduce

```powershell
{ Get-Command -ea Stop -Name ' '  } | Should -Throw -ErrorId CommandNotFoundException
```

## Expected behavior

The test should succeed; that is, the usual `CommandNotFoundException` should be reported.

## Actual behavior

The test fails, because the error mentioned above is reported instead.

```none
Expected an exception, with FullyQualifiedErrorId 'CommandNotFoundException' to be thrown, 
but the FullyQualifiedErrorId was 
'System.IndexOutOfRangeException,Microsoft.PowerShell.Commands.GetCommandCommand'
```

## Environment data

```none
PowerShell Core 7.1.0-preview.5
```

### PR DESCRIPTION
**Pull request by jakekerr**
# PR Summary

Fix #13342. Checks if _commandName is empty or null in the DoPowerShellRelativePathLookup() method. Removes a Dbg.Assert() that checks if the name is empty or null. The assertion fails because _commandName is trimmed in setupPathSearcher().

## PR Context

After this change `Get-Command ' '` will return a CommandNotFoundException instead of a IndexOutOfRangeException. The Dbg.Assert was removed because according to Issue #10165 whitespace should be a valid argument for Get-Command.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]
Please add test cases

**Comment by jackerr3** [comment]
I see now that an alternate and maybe cleaner solution would involve changing only this line `_commandName = _commandName.TrimEnd(Utils.Separators.PathSearchTrimEnd);` to create a new variable which is only used for the `setupPathSearcher()` routine. Creating a new variable for path searching may be more desirable than altering the existing `_commandName` string. If `_commandName` doesn't get trimmed we don't ever have to worry about the IndexOutOfRangeException being thrown.

File: src/System.Management.Automation/engine/CommandSearcher.cs [hunk 1]
```diff
@@ -1099,7 +1099,7 @@ private bool ShouldSkipCommandResolutionForConstrainedLanguage(CommandInfo? resu
                 // Relative Path:       ".\command.exe"
                 // Home Path:           "~\command.exe"
                 // Drive Relative Path: "\Users\User\AppData\Local\Temp\command.exe"
-                if (_commandName[0] == '.' || _commandName[0] == '~' || _commandName[0] == '\\')
+                if (!string.IsNullOrEmpty(_commandName) && (_commandName[0] == '.' || _commandName[0] == '~' || _commandName[0] == '\\'))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandSearcher.cs:1102]: `_commandName` is defined as non-nullable so we can do not check null.
Also it is on a hot path and we could optimize the code a bit. I suggest add `&& _commandName.Length != 0` in `if` above, define `char firstChar = _commandName[0]` and use the `firstChar` in the line.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1 [hunk 2]
```diff
@@ -101,6 +101,10 @@ Describe "Get-Command Feature tests" -Tag Feature {
         It "No results if wildcard is used" {
             Get-Command i-psd* -UseAbbreviationExpansion | Should -BeNullOrEmpty
         }
+
+        It "Returns CommandNotFoundException if a single space is used" {
+            {Get-Command ' ' -ErrorAction Stop} | Should -Throw -ErrorId "CommandNotFoundException,Microsoft.PowerShell.Commands.GetCommandCommand"
+        }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1:107]: For consistency we could add tests for null and empty arguments. Please use `-TestCases`.

**Comment by iSazonov** [comment]
> Creating a new variable for path searching may be more desirable than altering the existing _commandName string.

I think it makes no sense to create new variable - this complicates code and don't protect underlying _internal_ methods.

File: src/System.Management.Automation/engine/CommandSearcher.cs (hunk 1 above)
Comment (jackerr3) [reply src/System.Management.Automation/engine/CommandSearcher.cs:1102]: @iSazonov just to confirm. Is this along the lines of what you're suggesting? 
`                char firstChar = _commandName[0];`
                `if (_commandName.Length != 0 && (firstChar == '.' || firstChar == '~' || firstChar == '\\'))`

I agree it would be more optimized but the issue of this PR (`Get-Command ' '` returns the wrong exception) wouldn't be fixed. Sorry I may be misunderstanding. I'm still learning.

Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandSearcher.cs:1102]: > (!string.IsNullOrEmpty(_commandName)

You added the check to fix the issue. But `_commandName` is annotated as non-nullable and we can exclude null check and have `_commandName.Length != 0`. And the check should be in previous `if` condition (before `char firstChar = _commandName[0]`).

File: src/System.Management.Automation/engine/CommandSearcher.cs [hunk 3]
```diff
@@ -1099,13 +1099,19 @@ private bool ShouldSkipCommandResolutionForConstrainedLanguage(CommandInfo? resu
                 // Relative Path:       ".\command.exe"
                 // Home Path:           "~\command.exe"
                 // Drive Relative Path: "\Users\User\AppData\Local\Temp\command.exe"
-                if (!string.IsNullOrEmpty(_commandName) && (_commandName[0] == '.' || _commandName[0] == '~' || _commandName[0] == '\\'))
+
+                //char firstChar = _commandName[0];
+                if (_commandName.Length != 0)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandSearcher.cs]: Please remove commented code and move the condition in line 1104 to line 1090.

Comment (jackerr3) [reply src/System.Management.Automation/engine/CommandSearcher.cs]: Oh yeah thanks. Can't believe I missed that.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1 [hunk 4]
```diff
@@ -178,5 +179,15 @@ Describe "Get-Command" -Tag CI {
...
             $Result | Should -BeOfType [System.Management.Automation.CommandInfo]
         }
+
+        It "Throws <expected> exception if <name> name is used" -TestCases @(
+            @{ Name = ' '; expected = "CommandNotFoundException,Microsoft.PowerShell.Commands.GetCommandCommand" }
+            @{ Name = ''; expected = "ParameterArgumentValidationError,Microsoft.PowerShell.Commands.GetCommandCommand" }
+            @{ Name = $null; expected = "ParameterArgumentValidationError,Microsoft.PowerShell.Commands.GetCommandCommand" }
+        ) {
+            param($name, $expected)
+            $shortExpected = $expected -split ","
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1:189]: It is not used.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1 [hunk 5]
```diff
@@ -101,6 +101,7 @@ Describe "Get-Command Feature tests" -Tag Feature {
         It "No results if wildcard is used" {
             Get-Command i-psd* -UseAbbreviationExpansion | Should -BeNullOrEmpty
         }
+
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1:104]: Please remove extra empty line.
```suggestion
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1 [hunk 6]
```diff
@@ -178,5 +179,14 @@ Describe "Get-Command" -Tag CI {
 
             $Result | Should -BeOfType [System.Management.Automation.CommandInfo]
         }
+
+        It "Throws <expected> exception if <name> name is used" -TestCases @(
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1:183]:
```suggestion
        It "Throws '<expected>' exception if '<Name>' command name is used" -TestCases @(
```
Below please use Name (for the header) and Value (for test).

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1 [hunk 7]
```diff
@@ -178,5 +179,14 @@ Describe "Get-Command" -Tag CI {
...
+            @{ Name = ' '; expected = "CommandNotFoundException,Microsoft.PowerShell.Commands.GetCommandCommand" }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Command.Tests.ps1:184]: Like:
```suggestion
            @{ Name = 'space'; Value = ' '; expected = "CommandNotFoundException,Microsoft.PowerShell.Commands.GetCommandCommand" }
```

**Review by iSazonov — APPROVED** [review]

**Comment by jakekerr** [comment]
@SteveL-MSFT what do you think of the added test cases? Thanks.

**Comment by jakekerr** [comment]
@adityapatwardhan I don't mean to bother you but I just wanted to know if there's anything else I can do to help get this PR approved? Thanks.

**Comment by adityapatwardhan** [comment]
Currently working on some high priority items. I will be reviewing this after. Thank you for your patience.

**Comment by adityapatwardhan** [comment]
@SteveL-MSFT can you update your review

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by iSazonov** [comment]
@jakekerr Thanks for your contribution!
